### ISSUE TITLE
minio + Azure + Arq: "sha1 of uploaded file does not match" error

### ISSUE THREAD

**Issue by tomconte** [minio/minio#6550]
## Expected Behavior
Arq should work through the minio Azure gateway. Since it looks from previous issues that Arq was tested with minio before, this might be specific to the Azure gateway. There is no error from the logs (see below), so it seems that there is something that Arq is not expecting, but unfortunately not logging.

Would appreciate your input @krishnasrinivas @sreitshamer; I know this is a corner case but fixing this would allow me to use Arq with Azure Storage. 

## Current Behavior
When adding a folder to backup, an error "sha1 of uploaded file does not match" is received when Arq creates the encryption file.

## Steps to Reproduce (for bugs)
Start a minio instance with `minio gateway azure`. Use the minio endpoint to configure Arq.

## Your Environment

minio version:

```
Version: 2018-09-25T21:34:43Z
Release-Tag: RELEASE.2018-09-25T21-34-43Z
Commit-ID: aa4e2b1542b98097e08680f21b790de0b776378c
```

minio gateway runs on a macOS 10.11 machine.

Arq logs:

```
2018/10/01 17:57:41:439  DETAIL [thread 183eb] creating directory /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2
2018/10/01 17:57:41:439  DETAIL [thread 183eb] creating directory s3:10.0.0.100:/testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2
2018/10/01 17:57:41:443  DEBUG [thread 183eb] 0x7fa1f32379a0 RemoteFS.m:343 cache not loaded; getting items for parent dir of /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2/encryptionv3.dat
2018/10/01 17:57:41:443  DEBUG [thread 183eb] 0x7fa1f32379a0 RemoteFS.m:372 cachedItemsByNameForDirectory returned nil; error=Error Domain=ItemsDBErrorDomain Code=-31 "itemsByNameInDirectory: cache not loaded for directory /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2" UserInfo={NSLocalizedDescription=itemsByNameInDirectory: cache not loaded for directory /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2}
2018/10/01 17:57:41:443  DETAIL [thread 183eb] getting items in directory s3:10.0.0.100:/testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2
2018/10/01 17:57:41:443  DEBUG [thread 183eb] 0x7fa1f3071710 URLConnection.m:46 GET http://10.0.0.100:9000/testminiotco-arq/?prefix=C823368A-2017-4A7A-89A4-A4669AC094D2/&delimiter=/&max-keys=500
2018/10/01 17:57:41:444  DEBUG [thread 183eb] 0x7fa1f3071710 URLConnection.m:147 NSURLConnection started with no request body
2018/10/01 17:57:41:606  DEBUG [thread 183eb] 0x7fa1f3071710 URLConnection.m:193 response: status = 200, Content-Length = (null), Transfer-Encoding = Identity
2018/10/01 17:57:41:606  DEBUG [thread 183eb] 0x7fa1f3071710 URLConnection.m:209 response body with no Transfer-Encoding; responseData is 302 bytes
2018/10/01 17:57:41:606  DEBUG [thread 183eb] 0x7fa1f2c048b0 S3Request.m:257 HTTP 200; returning response length=302
2018/10/01 17:57:41:607  DEBUG [thread 183eb] 0x7fa1f2d08ef0 TargetItemsDB.m:369 inserting 0 rows into the cache db for /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2
2018/10/01 17:57:41:608  DEBUG [thread 183eb] 0x7fa1f2d08ef0 TargetItemsDB.m:375 inserted 0 rows into the cache db for /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2
2018/10/01 17:57:41:609  DEBUG [thread 183eb] 0x7fa1f32379a0 RemoteFS.m:352 didn't find encryptionv3.dat in parent dir /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2
2018/10/01 17:57:41:609  DETAIL [thread 183eb] creating file s3:10.0.0.100:/testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2/encryptionv3.dat
2018/10/01 17:57:41:609  DEBUG [thread 183eb] 0x7fa1f2fbc1e0 URLConnection.m:46 PUT http://10.0.0.100:9000/testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2/encryptionv3.dat
2018/10/01 17:57:41:609  DEBUG [thread 183eb] 0x7fa1f2fbc1e0 URLConnection.m:145 NSURLConnection started with request body (180 bytes)
2018/10/01 17:57:41:689  DEBUG [thread 183eb] 0x7fa1f2fbc1e0 URLConnection.m:193 response: status = 200, Content-Length = 0, Transfer-Encoding = (null)
2018/10/01 17:57:41:689  DEBUG [thread 183eb] 0x7fa1f2fbc1e0 URLConnection.m:209 response body with no Transfer-Encoding; responseData is 0 bytes
2018/10/01 17:57:41:689  DEBUG [thread 183eb] 0x7fa1f337d4e0 S3Request.m:257 HTTP 200; returning response length=0
2018/10/01 17:57:41:689  DEBUG [thread 183eb] 0x7fa1f32e67d0 S3Service.m:354 uploaded 180 bytes to /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2/encryptionv3.dat in 0.1 seconds (2249 bytes/sec)
2018/10/01 17:57:41:690  DEBUG [thread 1afab] 0x7fa1f32e8f00 HTTPInputStream.m:30 deallocing HTTPInputStream 0x7fa1f32e8f00
```

Last entries from `minio.log`:

```
[REQUEST (objectAPIHandlers).PutObjectHandler-fm] [153840946.161095] [2018-10-01 17:57:41 +0200]
PUT /testminiotco-arq/C823368A-2017-4A7A-89A4-A4669AC094D2/encryptionv3.dat
Host: 10.0.0.100
Content-Length: 180
Authorization: AWS4-HMAC-SHA256 Credential=testminiotco/20181001/us-east-1/s3/aws4_request,SignedHeaders=content-length;host;x-amz-content-sha256;x-amz-date,Signature=75b7e4cba5fb8ed400cc18e84a33c0b6717a577cfd5a094fc6cf8e0248559689
User-Agent: Arq/5.14 CFNetwork/760.9 Darwin/15.6.0 (x86_64)
Connection: keep-alive
Accept: */*
Accept-Language: en-us
X-Amz-Date: 20181001T155741Z
Accept-Encoding: gzip, deflate
X-Amz-Content-Sha256: ecdc64e494481a868f469ebabdb7b9f0be8272b2b80a30d4de122ef5a6b2d54d

<BODY>

[RESPONSE] [153840946.161095] [2018-10-01 17:57:41 +0200]
200 OK
X-Xss-Protection: 1; mode=block
Content-Security-Policy: block-all-mixed-content
X-Amz-Request-Id: 1559874126681654
Etag: "0x8D627B69E804C5A-1"
Server: Minio/RELEASE.2018-09-25T21-34-43Z (darwin; amd64)
Accept-Ranges: bytes
Vary: Origin

<BODY>
```

**Comment by arqbackup** [comment]
The Etag response header is supposed to be the MD5 hash of the request body. https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html

**Comment by harshavardhana** [comment]
No it is not true always - a multipart object doesn't have md5sum of the entire object, etag shouldn't be assumed to be md5sum. 

This is why Amazon states ETag is an opaque value 

For proper definition look here 

https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html

> The entity tag is a hash of the object. The ETag reflects changes only to the contents of an object, not its metadata. **The ETag may or may not be an MD5 digest of the object data. Whether or not it is depends on how the object was created and how it is encrypted as described below:**

**Comment by arqbackup** [comment]
Well, if you don't upload multipart, it's supposed to be the MD5 hash. So, minio doesn't behave the same way as S3, right?

**Comment by harshavardhana** [comment]
What I mean is you are assuming a whole lot here,

AWS SDKs by default upload all files bigger than 5MiB, s3cmd uploads all files bigger than 15MiB and mc uploads all files bigger than 64MiB as multipart, objects uploaded through SSE-C they don't need to be multipart. Objects encrypted after a CopyObject server side.

The question is does Arq ignore these objects ? If so how ? The mechanism is wrong to assume ETag was md5sum when the nature is not guaranteed on what happens to an object.

For a better implementation to honor these assumptions Rclone, restic are better examples which uploads extra information necessary for future validation of the object work completely fine with azure gateway.

**Comment by arqbackup** [comment]
Arq doesn't do any of those things. It just does PUTs, not multipart.
I don't understand the confusion. If I do a simple PUT to S3, I get an MD5 hash in the Etag. It's been that way since the first day of S3. What is minio returning in the Etag?

**Comment by tomconte** [comment]
It looks like the gateway is returning the value of the Azure Etag property in the response?

https://github.com/minio/minio/blob/master/cmd/gateway/azure/gateway-azure.go#L691

Unfortunately in the [Azure Blob API](https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob-properties), the Etag is not equal to the MD5 hash of the object ("The ETag contains a value that you can use to perform operations conditionally").

It looks like minio [adds the "-1" suffix](https://github.com/minio/minio/blob/1111419d4aea8356a4a21cea7ba93560b42c87ae/cmd/utils.go#L304) to "indicate ETag is not a checksum" but I don't know if this behavior is part of the S3 protocol definition.

**Comment by harshavardhana** [comment]
-1 is added to indicate it's a multipart object from S3 protocol point of view. Some clients such as s3cmd ignore the value treating it as multipart operation.

**Comment by tomconte** [comment]
The doc for [S3 PUT](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html) does seem to state clearly "you can calculate the MD5 while putting an object to Amazon S3 and compare the returned ETag to the calculated MD5 value" though?

**Comment by arqbackup** [comment]
If minio is supposed to be S3-compatible and it's acting as a gateway to Azure, should it perhaps calculate the MD5 hash itself and return it in the response for a simple S3 PUT? That would make it 100% S3-compatible for that API call.

**Comment by harshavardhana** [comment]
> The doc for [S3 PUT](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html) does seem to state clearly "you can calculate the MD5 while putting an object to Amazon S3 and compare the returned ETag to the calculated MD5 value" though?

Yes, indeed that statement signifies that but building an application around it is another thing making it too much S3 centric, which is what @arqbackup did which is unfortunate since it doesn't leverage multipart and other APIs which have other benefits i.e faster uploads etc. For example, SSE-C is not supported as well assuming that has never come up as a requirement for backups.  That is a design decision and outside the scope here let's leave that aside. 

> If minio is supposed to be S3-compatible and it's acting as a gateway to Azure, should it perhaps calculate the MD5 hash itself and return it in the response for a simple S3 PUT? That would make it 100% S3-compatible for that API call.

It is we already validate md5sum anyways if the Content-Md5 is set, the problem is preserving the value for later operations and applications such as Arq assuming this to be present and persisted. This is the reason that we return what is calculated by the backend.

**Comment by arqbackup** [comment]
I'm sorry but I don't understand what is "unfortunate". Arq uses multiple threads to upload in parallel, so it doesn't need multipart upload, and it encrypts data client-side itself so it doesn't need SSE-C. None of these things you're talking about are relevant to whether or not minio is 100% S3-compatible.

The problem is really clear. Minio doesn't return the same response headers as S3. So, it's not 100% S3-compatible.

It sounds like the answer is no, minio won't be made 100% S3-compatible, at least in the near term. So I guess it would be up to me to modify Arq to be Minio-compatible.

**Comment by arqbackup** [comment]
FWIW Arq works fine with minio when it's storing the data itself (not acting as a gateway to azure). Arq also works fine with many other S3-compatible providers like Wasabi, DigitalOcean Spaces, DreamObjects. They all return the correct Etag value.

**Comment by harshavardhana** [comment]
> It sounds like the answer is no, minio won't be made 100% S3-compatible, at least in the near term. So I guess it would be up to me to modify Arq to be Minio-compatible.

For now, it's not possible we will have to work backward and re-evaluate this situation. Will update here on what we decide and if it's technically possible even.  

> FWIW Arq works fine with minio when it's storing the data itself (not acting as a gateway to azure). Arq also works fine with many other S3-compatible providers like Wasabi, DigitalOcean Spaces, DreamObjects. They all return the correct Etag value.

Yep, that is what has been tested, we were not sure ETag is a strong requirement. Gateway is also unique in Minio So I guess these compatibility issues will come up.

**Comment by kannappanr** [comment]
@arqbackup The problem is that in this case, the E-Tag is generated by Azure, which doesn't have to be s3-compatible. If the same test were to be done on an s3 backend, instead of Azure, you will not see this issue.

Also,  there are other scenarios where multipart is involved, E-Tag will not equal to MD5 sum. Same is true with SSE-C.  You can probably make sure that your application does only single part upload and that would mean that you support backup of objects with size upto 5GB (https://docs.aws.amazon.com/AmazonS3/latest/dev/UploadingObjects.html) For objects beyond 5GB, your theory doesn't hold good.

**Comment by harshavardhana** [comment]
@kannappanr Yes I think @arqbackup understands this situation, Arq software is written with a different design in mind. It splits the data upfront such that the object uploaded is not really a namespace of actual users content instead of its what Arq understands.

So each of these PUT's are tracked with Etag. We can discuss this through and re-evaluate our decisions to see if ETag can indeed be preserved on the cloud providers by adding some extra data which is Minio specific.

**Comment by harshavardhana** [comment]
Moving this as blocked for discussion

**Comment by harshavardhana** [comment]
Sent a PR to fix this situation #6552 @tomconte @arqbackup

**Comment by tomconte** [comment]
I am testing a build from the PR branch and I was able to start the backup process. 
Looks good! Many thanks to everyone 🙌🏻
